Math/NumberTheory/Recurrencies/Linear.hs

```diff
@@ -20,6 +21,20 @@ module Math.NumberTheory.Recurrencies.Linear
 #include "MachDeps.h"
 
 import Data.Bits
+import Numeric.Natural
+
+#if MIN_VERSION_base(4,8,0)
+#else
+import Data.Word
+#endif
+
+-- | Infinite sequence of factorials, starting with 0!
+factorial :: (Num a, Enum a) => [a]
```

Comment:
I'm not sure if This has the memory behavior you want. Namely the specializations will be shared across usages. But polymorphic one is a function from the type class dictionaries to the lazy sequence.